pandas/core/frame.py

```diff
@@ -5007,6 +5007,121 @@ def rename(
             errors=errors,
         )
 
+    @overload
```

Comment:
I was told to add them directly to these files since existing overloads were already there

Reply:
@jreback yes, that would be great! Thing is though, the `.pyi` files require you to define _all_ methods of a module

Given the sheer number of methods this module has, I'd suggest taking this PR with the overloads here, and then moving the overloads (along with annotations for all other methods) to a `pandas/core/frame.pyi` file

Reply:
@LarWong I'll get back to you on typing `value`

Reply:
> can we not put these in .pyi?

I don't think we want to go the route of using stubs for python files

> @jreback yes, that would be great! Thing is though, the `.pyi` files require you to define _all_ methods of a module

not that I think it should be done here, but it is possible to partially type a module using stubs.

https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#incomplete-stubs

> Partial modules (i.e. modules that are missing some or all classes, functions, or attributes) must include a top-level __getattr__() function marked with an # incomplete comment (see example below).

Reply:
Thanks @simonjayhawkins , didn't know that was possible. Why not use partial stubs for overloads? Because some methods, like `Series.drop`

https://github.com/pandas-dev/pandas/blob/84d9c5ed7965b3cc667b5a3a61d5911b1748af49/pandas/core/series.py#L4478-L4487

have 5 (five!) arguments with defaults before `inplace`, leading to...34 overloads :exploding_head: ! And even if we disallowed `labels` being passed as `None` in that one, that would still leave us with 18 overloads!

---

For typing `value` here, do you think `Scalar | Mapping[Hashable, Scalar] | Series | DataFrame | None` would be correct?

Reply:
The problem is that if a stub file is used it takes precedence over the python file. so we cannot ensure internal consistency and need to duplicate the type annotations to be able to check the functions in the module itself.

Reply:
> have 5 (five!) arguments with defaults before `inplace`, leading to...34 overloads

I think we are planning to to drop the inplace argument, hopefully pandas 2.0 and we won't need all these overloads. https://github.com/pandas-dev/pandas/issues/16529